### PR TITLE
fix(ocpi-2.2.1): Fixed usage of tokens & review PlatformRepository

### PR DESCRIPTION
**Pull request by lilgallon**
## Change notes

**only for 2.2.1**

**Credentials logic**

The credentials logic completely changed. Token B and Token C were replaced by Client Token and Server Token. The Token A usage stays the same.

The Credentials Token A is used by the sender to communicate with the receiver (only when initiating registration). After registration, it is invalidated.

Then, there are two tokens, and the token to use is specified by the registration process.
 - When you are the receiver during registration:
   - OCPI Token B is the client token, the one you have to use to send requests
   - OCPI Token C is the server token, the one you have to use to check if requests are properly authenticated
 - When you are the sender during registration:
   - OCPI Token B is the server token, the one you have to use to check if requests are properly authenticated
   - OCPI Token C is the client token, the one you have to use to send requests

This is why we do not use OCPI's token B and token C naming. According to who you are in the registration process, you have to use a different token. Using "Client Token" and "Server Token" simplifies that process of picking the right token for the right operation.

Before these changes, the library would not work when used while being the receiver of the registration process.

**PlatformRepository modifications**

Unused methods are removed. Some were renamed and their signature updated to match the new credentials logic. The unregistration is now in one unique method, rather than 5 different methods.
 
## Migration guide

It's a breaking change, it also breaks your `Platform` objects.

You will have to update your `PlatformRepository` implementation to match the new interface. Everything is documented in the file. You have to re-register with your partners. Or you can mnually update your `Platform` object to reflect the changes (clientToken / serverToken instead of tokenB / tokenC).

## Notes

This will allow us to bump the lib to 0.1.0 as it should now work properly, and it's now usable (not production ready yet obvisouly).

### CODE REVIEW

File: ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt [hunk 1]
```diff
@@ -70,45 +71,53 @@ fun String.decodeBase64(): String = Base64.getDecoder().decode(this).decodeToStr
...
-suspend fun PlatformRepository.buildAuthorizationHeader(baseUrl: String, allowTokenAOrTokenB: Boolean = false) =
-    if (allowTokenAOrTokenB) {
-        getCredentialsTokenC(platformUrl = baseUrl)
-            ?: getCredentialsTokenB(platformUrl = baseUrl)
-            ?: getCredentialsTokenA(platformUrl = baseUrl)
-            ?: throw throw OcpiClientGenericException(
-                "Could not find CREDENTIALS TOKEN A OR B OR C associated with platform $baseUrl"
+suspend fun PlatformRepository.buildAuthorizationHeader(
+    platformUrl: String,
+    allowTokenA: Boolean = false
```
Comment (xhanin) [inline ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt:83]: I wonder if  we want to "allow" the use of tokenA, or actually force to use it? The only case of using tokenA is when we initialize the credentials handshake, where we call 3 endpoints with tokenA (and in that case we are sure we want to use tokenA). Then we will always use the client token. I will go on with the review to see if what I say makes sense or not.

File: ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt [hunk 2]
```diff
@@ -243,27 +253,41 @@ fun HttpRequest.parseAuthorizationHeader() = getHeader(Header.AUTHORIZATION)
...
- * - tokenB: Valid in sender context, during sender registration (only for receiver -> sender calls)
- * - tokenC: Valid when the sender is registered with the receiver (only for sender -> receiver)
- *
  * @throws OcpiClientInvalidParametersException if the token is invalid, otherwise does nothing
  * @throws OcpiClientNotEnoughInformationException if the token is missing
  * @throws HttpException if the authorization header is missing
+ *
  */
-suspend fun PlatformRepository.tokenFilter(httpRequest: HttpRequest) {
+suspend fun PlatformRepository.tokenFilter(
```
Comment (xhanin) [inline ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt:261]: I would rename that to `checkToken`: we are not building a filter, we really checking the provided token

File: ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt [hunk 3]
```diff
@@ -243,27 +253,41 @@ fun HttpRequest.parseAuthorizationHeader() = getHeader(Header.AUTHORIZATION)
...
+     * From OCPI 2.2.1 doc:
+     * When a server receives a request with a valid CREDENTIALS_TOKEN_A, on another module than: credentials or
+     * versions, the server SHALL respond with an HTTP 401 - Unauthorized status code.
+     *
+     * So, we allow token A only if we are in this case.
+     */
+
+    val allowTokenA = httpRequest.path.contains("versions") ||
+        httpRequest.path.contains("/{versionNumber}") ||
+        httpRequest.path.contains("credentials")
```
Comment (xhanin) [inline ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt:276]: I would rather use a dedicated parameter to tell if we are in the list of endpoints allowed in the handshake or not, to allow tokenA only there. I find this detection mechanism error prone, and as this is a security feature I think it's worth to make it more strict

File: ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt [hunk 4]
```diff
@@ -243,27 +253,41 @@ fun HttpRequest.parseAuthorizationHeader() = getHeader(Header.AUTHORIZATION)
...
+
+    if (!validToken) {
+        throw OcpiClientInvalidParametersException("Invalid server token (token A allowed: $allowTokenA): $token")
     }
 }
 
-suspend fun TransportClientBuilder.buildFor(module: ModuleID, platform: String, platformRepository: PlatformRepository) =
+suspend fun TransportClientBuilder.buildFor(
+    module: ModuleID,
+    platform: String,
```
Comment (xhanin) [inline ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt:288]: this string is misleading: it's the platformUrl we need here. That's something that is not fully obvious to e yet that we use this platformUrl everywhere to identifiy the platform, I wonder if we could rather use a platform id or `partnerPlatformId`. But this goes beyond the scope of this review

**Review by xhanin — APPROVED** [review]
I have a few comments, but they are beyond the scope of this PR, and could be discussed / handled in another PR.
Therefore this PR is good for me as is, and the updates for the other versions can be handled separately in other PR / issues.

File: ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt (hunk 1 above)
Comment (lilgallon) [reply ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt:83]: created issue: https://github.com/IZIVIA/ocpi-toolkit/issues/19

File: ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt (hunk 2 above)
Comment (lilgallon) [reply ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt:261]: fixed

File: ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt (hunk 3 above)
Comment (lilgallon) [reply ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt:276]: issue created: https://github.com/IZIVIA/ocpi-toolkit/issues/20

File: ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt (hunk 4 above)
Comment (lilgallon) [reply ocpi-toolkit-2.2.1/src/main/kotlin/com/izivia/ocpi/toolkit/common/HttpUtils.kt:288]: fixed

**Pull request merged**